force-app/main/default/lwc/compositionContactSearch/__tests__/compositionContactSearch.test.js

```diff
@@ -0,0 +1,106 @@
+import { createElement } from 'lwc';
+import CompositionContactSearch from 'c/compositionContactSearch';
+import findContacts from '@salesforce/apex/ContactController.findContacts';
+
+const mockFindContacts = jest.mock(
+    '@salesforce/apex/ContactController.findContacts',
+    () => {
+        return {
+            default: jest.fn()
+        };
+    },
+    { virtual: true }
+);
+
+describe('c-composition-contact-search', () => {
+    afterEach(() => {
+        // The jsdom instance is shared across test cases in a single file so reset the DOM
+        while (document.body.firstChild) {
+            document.body.removeChild(document.body.firstChild);
+        }
+    });
+
+    it('renders one lightning-input field labelled "Search"', () => {
+        // Create initial element
+        const element = createElement('c-composition-contact-search', {
+            is: CompositionContactSearch
+        });
+        document.body.appendChild(element);
+
+        // Query lightning-input field elements
+        const inputFieldEls = element.shadowRoot.querySelectorAll(
+            'lightning-input'
+        );
+
+        // Check length of rendered lightning-input elements
+        expect(inputFieldEls.length).toBe(1);
+
+        // Check label of lightning-input element
+        expect(inputFieldEls[0].label).toBe('Search');
+    });
+
+    it('renders one contact tile based on user input', () => {
+        mockFindContacts.mockResolvedValue = [{ Id: '99', Name: 'Amy Taylor' }];
+
+        // Create initial element
+        const element = createElement('c-composition-contact-search', {
+            is: CompositionContactSearch
+        });
+        document.body.appendChild(element);
+
+        // Query lightning-input field element
+        const inputFieldEl = element.shadowRoot.querySelector(
+            'lightning-input'
+        );
+        inputFieldEl.value = 'Amy';
+        inputFieldEl.dispatchEvent(new CustomEvent('change'));
+
+        jest.runAllTimers();
+
+        // Return a promise to wait for any asynchronous DOM updates. Jest
+        // will automatically wait for the Promise chain to complete before
+        // ending the test and fail the test if the promise ends in the
+        // rejected state
+        return Promise.resolve().then(() => {
+            // TODO rw: check for method call of findContacts
```

Comment:
@trevor-bliss @kv here I'm not getting return data from imperative Apex. I guess I'm doing the data mock further up wrong.

Reply:
I don't think the user "KV" is going to know the answer to this. @kevinv11n is probably interested though 😛

There's a lot of options and knobs to turn when it comes to mocking. Here's how I got it working:

// at top of test file - sets up initial mock to be shared between 
```
import findContacts from '@salesforce/apex/ContactController.findContacts';
jest.mock(
    '@salesforce/apex/ContactController.findContacts',
    () => {
        return {
            default: jest.fn()
        };
    },
    { virtual: true }
);
```

// in test case - sets up mock return value 
```
findContacts.mockResolvedValue([{ Id: '99', Name: 'Amy Taylor' }]);
```

// in test case - verify mock call
```
expect(findContacts).toHaveBeenCalledWith({ ...whatever expected shape is... });
```

Reply:
I applied this and it's throwing in the test run an error:

```
    TypeError: Cannot read property 'then' of undefined

      16 |         // eslint-disable-next-line @lwc/lwc/no-async-operation
      17 |         this.delayTimeout = setTimeout(() => {
    > 18 |             findContacts({ searchKey })
```

Reply:
You need to change `findContacts.mockResolvedValue = APEX_CONTACTS_SUCCESS;` to `findContacts.mockResolvedValue(APEX_CONTACTS_SUCCESS);`. 

The test will then fail with the message:
```
[{"Id": "99", "Name": "Amy Taylor"}]
    as argument 1, but it was called with
      {"searchKey": "Amy"}.
```

This is expected though if you look at how `findContacts` is called in `compositionContactSearch.js`. Based on the title "renders one contact tile based on user input", I don't think you should be verifying what `findContacts` is called with. That should be a separate test case.

Another thing to consider is that you'll need to wait an extra microtask for the DOM to rerender. The first microtask for the `findContacts` promise to resolve and set the `contacts` property, the second for the rerender.

Reply:
Updated. Although what's the recommendation for testing the `catch`? I tried `rejected`, but the c-error-panel doesn't get rendered.

https://github.com/trailheadapps/lwc-recipes/pull/33/files#diff-3a162ea4596850fe3ca4f34927482727R89